rir/src/compiler/opt/type_speculation.cpp

```diff
@@ -91,25 +91,32 @@ void TypeSpeculation::apply(RirCompiler&, ClosureVersion* function,
         if (!speculate.count(cp))
             return;
 
-        bb = bb->trueBranch();
-
         for (auto sp : speculate[cp]) {
-            auto i = sp.first;
-
-            auto ip = bb->begin();
-            if (i->bb() == bb)
-                ip = bb->atPosition(i) + 1;
-
+            auto speculationTarget = sp.first;
             TypeTest::Info& info = sp.second;
+            // The instruction requiring speculation could be before or after
+            // the cp, so we need to find the place that is after both to insert
+            // the typetest and the assume
+            BB* targetBB;
+            BB::Instrs::iterator insertPosition;
+            targetBB = bb->trueBranch();
+            if (speculationTarget->bb()->beforeInCfg(targetBB)) {
+                insertPosition = targetBB->begin();
+            } else {
+                if (targetBB->beforeInCfg(speculationTarget->bb()))
+                    targetBB = speculationTarget->bb();
+                insertPosition = targetBB->atPosition(speculationTarget) + 1;
+            }
```

Comment:
this does not seem right. how do you know that the checkpoint is still available at that position?

Reply:
because the previous code calculates exactly which is the cp for that instruction?

Reply:
ok, I think I get now what you are trying to do.

on line 66 you know if we have the  `next` or the `at` checkpoint. so no need to do this dance here to figure out again. if we did `checkpoint.at`, then we must insert the assume right after i (like we do now). if we did `checkpoint.next` then we must insert the assume at `cp->next()->begin()`. right?

so it seems like this was somehow completely broken, but never triggered any issues? do you understand why?

Reply:
I've seen similar things when working around the more aggressive delaying of instructions. Also, the environment elision had some issues. My impression is that we were performing all these speculative optimizations only in very few and simple places...